### DISCUSSION TITLE
I cannot get (publish?) my repository into GitHub

### DISCUSSION THREAD

**Discussion by SimpleSamples** [community/community#22287] [New to GitHub]
<p>I have been programming for nearly half a century but I still am highly lost with Git and GitHub. The following is a log from Visual Studio Code.</p>
<pre><code>Looking for git in: C:\Program Files\Git\bin\git.exe
Using git 2.17.1.windows.2 from C:\Program Files\Git\bin\git.exe
&gt; git rev-parse --show-toplevel
&gt; git config --get commit.template
Open repository: g:\Sam\Documents\Source\Repos
&gt; git status -z -u
&gt; git symbolic-ref --short HEAD
&gt; git rev-parse master
&gt; git rev-parse --symbolic-full-name master@{u}
fatal: no upstream configured for branch 'master'
&gt; git for-each-ref --format %(refname) %(objectname) --sort -committerdate
&gt; git remote --verbose
&gt; git check-ignore -z --stdin
&gt; git status -z -u
&gt; git symbolic-ref --short HEAD
&gt; git rev-parse master
&gt; git rev-parse --symbolic-full-name master@{u}
fatal: no upstream configured for branch 'master'
&gt; git for-each-ref --format %(refname) %(objectname) --sort -committerdate
&gt; git remote --verbose
&gt; git check-ignore -z --stdin
&gt; git add -A -- .
&gt; git commit --quiet --allow-empty-message --file - --all
&gt; git config --get-all user.name
&gt; git config --get-all user.email
&gt; git status -z -u
&gt; git symbolic-ref --short HEAD
&gt; git rev-parse master
&gt; git rev-parse --symbolic-full-name master@{u}
fatal: no upstream configured for branch 'master'
&gt; git for-each-ref --format %(refname) %(objectname) --sort -committerdate
&gt; git remote --verbose
</code></pre>
<p>Probably the problem is that I had a phantom repository within my main repository so I deleted it (using Visual Studio I am nearly certain, not the file system) and separately deleted it using GitHub online.</p>
<p>So let me back up to the beginning. I have been trying to use Visual Studio (not VS Code) to get a Web App project uploaded (published?) into GitHub. I usually must stumble around but I eventualy get it to work. This time I could not. So I tried GitHub Desktop but I could not get it to work either, it just gives generic errors too. So I tried VS Code and that is what the log is from. I searched for:</p>
<pre><code>"fatal: no upstream configured for branch" 'master'
</code></pre>
<p>And found another question that the answer probably does not apply here.</p>
<p>I found the following in StackOverflow:</p>
<ol>
<li><a href="https://stackoverflow.com/questions/45406162/visual-studio-git-not-updating-and-commit-issues" rel="nofollow noopener">Visual Studio Git Not Updating and Commit Issues</a></li>
<li><a href="https://stackoverflow.com/questions/51549407/i-cannot-publish-my-vs-code-repository-to-github" rel="nofollow noopener">git - I cannot publish my VS Code repository to GitHUB</a></li>
</ol>
<p>If either of those applies to my situation then I apologize for not knowing enough to understand them.</p>
<p>It is very frustrating for me that develpers of GUI interfaces to GitHub are unable to create an interface that is independent of the command-line commands and procedures. When I finally understand Git well enough I will create a GUI the way it should be done. At the moment I just want to uplaod some sample code for a totally different question.</p>

**Comment by mpboom** [comment]
<p>Hi <a class="mention" href="/simplesamples">@simplesamples</a> ,</p>
<p>No problem if you don’t understand it - we are here to help.</p>
<p>What exactly are you trying to achieve? If you’d just like to push (that is, I think, the most correct term to use) your local  <strong>master</strong> branch to the  <strong>master </strong> branch on GitHub you could use the following command:</p>
<pre><code>git push --set-upstream origin master
</code></pre>
<p>If the GH remote is called  <strong>origin</strong> (which it by default is), this will push your local  <strong>master </strong> to the  <strong>master</strong> on GitHub. You do need to execute this command manually, but after that automatically pushing using VS or VS Code should work fine without any manual actions from your side.</p>
<p>Let us know if you get more errors or need additional assistance!</p>

**Comment by SimpleSamples** [comment]
<p>Thank you, Mark.</p>
<p>When I use a command prompt and go to the repository’s path and issue that command I get:</p>
<pre><code>fatal: 'origin' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
</code></pre>
<p>So I am not sure what is happening. Since I am not experienced with the use of the command line, I am not sure where the credentials are coming from. I tried to figure that out and just doing that makes my head spin; there are so many ways to specify credentials. Well I think that since I use Visual Studio I should assume I am using whatever it uses. Anyway, to attempt to learn things I tried the following.</p>
<pre><code>G:\Sam\Documents\Source\Repos&gt;git config -l --global
user.name=Sam Hobbs
user.email=SimpleSamples@users.noreply.github.com
gui.recentrepo=C:/Users/Sam/source/repos
filter.lfs.clean=git-lfs clean -- %f
filter.lfs.smudge=git-lfs smudge -- %f
filter.lfs.process=git-lfs filter-process
filter.lfs.required=true
credential.helper=manager
</code></pre>
<p>And that is very interesting. The most interesting part is the recentrepo. Windows allows us to specify a different partition for our “Documents” folder but it does not support it very well. Ugh! The relevant repository is in my G: (as shown in the preceding) but the preceding says C: for the recentrepo. So that is a likely problem. Right? Does that provide enough for you to suggest what to try next?</p>

**Comment by mpboom** [comment]
<p>Hi <a class="mention" href="/simplesamples">@simplesamples</a> ,</p>
<p>Can you post the output of  <strong>git remote -v</strong> when inside the repository here? Thanks!</p>
<p><strong>Please check the output for sensitive information, and if so, censor it!</strong></p>

**Comment by SimpleSamples** [comment]
<p>I get nothing. I tried using both Git Bash and a Windows command prompt. Both produced no output, not even an error message.</p>
<p>I did that in the folder that has a .git directory; it has about eight projects in it.</p>
<p>So just to see if I could get that to work I went to my local repository for <a href="https://samhobbs.herokuapp.com/" rel="nofollow noopener">First Heroku Site</a>. There is not much there but the output of the specified command is:</p>
<pre><code>origin https://github.com/SamHobbsOrg/HerokuFirst.git (fetch)
origin https://github.com/SamHobbsOrg/HerokuFirst.git (push)
</code></pre>

**Comment by mpboom** [comment]
<p><a class="mention" href="/simplesamples">@simplesamples</a>  it is correct if Git then returns an empty response - it simply means you don’t have any remotes </p>
<p>You simply need to add your own Git repository as a remote (using this article: <a href="https://help.github.com/en/articles/adding-a-remote" rel="nofollow noopener">https://help.github.com/en/articles/adding-a-remote</a>) and then you can retry the command mentioned earlier.</p>

**Comment by SimpleSamples** [comment]
<p>Thank you <a class="mention" href="/mpboom">@mpboom</a>. I think I am making progress. After adding the remote I tried the push again and I get:</p>
<pre><code>G:\Sam\Documents\Source\Repos&gt;git push --set-upstream origin master
git: 'credential---help' is not a git command. See 'git --help'.
To https://github.com/SimpleSamples/Repos.git
 ! [rejected] master -&gt; master (fetch first)
error: failed to push some refs to 'https://github.com/SimpleSamples/Repos.git'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
</code></pre>
<p>I think the part about <em>credential—help</em> is my mistake resulting from my attempt to figure things out. I hope it is doing nothing except produce that warning. As ar as <em>you do not have locally</em> that gets back to my original post, I did delete a project that I know I should not have. I deleted it both in my local and remote repositories but I should have known that Git has data about things like that. So I need to clean up what is in the remote.</p>
<p>Would it work to just create a new remote? I have everything locally. Can I first rename the existing remote? Or I should delete the remote but if nothing else I do not mind using a different name.</p>

**Comment by SimpleSamples** [comment]
<p>I posted a reply but then I edited it a couple of times then it needed approval but I think now it is gone. I apologize if the following is a duplicate.</p>
<p>Thank you <a class="mention" href="/mpboom">@mpboom</a>. I think I am making progress. After adding the remote I tried the push again and I got:</p>
<pre><code>G:\Sam\Documents\Source\Repos&gt;git push --set-upstream origin master
git: 'credential---help' is not a git command. See 'git --help'.
To https://github.com/SimpleSamples/Repos.git
! [rejected] master -&gt; master (fetch first)
error: failed to push some refs to 'https://github.com/SimpleSamples/Repos.git'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
</code></pre>
<p>I think the part about <em>credential—help</em> is my mistake resulting from my attempt to figure things out. I hope it is doing nothing except produce that warning. As far <em>as you do not have locally</em> that gets back to my original post, I did delete a project that I know I should not have. I deleted it both in my local and remote repositories but I should have known that Git has data about things like that. So I need to clean up what is in the remote.</p>
<p>Would it work to just create a new remote? I have everything locally. Can I first rename the existing remote? Or I can delete the remote but if nothing else I do not mind using a different name.</p>

**Comment by mpboom** [comment, marked as answer]
<p><a class="mention" href="/simplesamples">@simplesamples</a> you probably made some commits to the repository online that aren’t on your computer yet. But that is no problem! Git is designed exactly for this.</p>
<p>Do a  <strong>git pull</strong> inside the repository and Git will “download” the commits you made online to your computer. If you have happend to have the same file on the GitHub site  <strong>and</strong> locally, you might get a  <strong>merge conflict</strong>. Even that is no big deal, just follow the instructions Git gives you and/or use this article: <a href="https://help.github.com/en/articles/about-merge-conflicts" rel="nofollow noopener">https://help.github.com/en/articles/about-merge-conflicts</a> .</p>
<blockquote>
<p>I think I am making progress.</p>
</blockquote>
<blockquote>
<p><strong>You definitely are!</strong></p>
</blockquote>
<blockquote>
<p> </p>
</blockquote>
<blockquote>
<p>Would it work to just create a new remote? I have everything locally. Can I first rename the existing remote? Or I can delete the remote but if nothing else I do not mind using a different name.</p>
</blockquote>
<blockquote>
<p><strong>No, changing the remotes won’t fix it. It’s just that you’ve made commits both locally and online. That leads to Git getting a bit out of sync - but remember: it was designed for this!</strong></p>
</blockquote>

**Comment by SimpleSamples** [comment]
<p>Hopefully this is just a progress report. I am getting:</p>
<pre><code>fatal: refusing to merge unrelated histories
</code></pre>
<p>If that is expected for the situation described previously then I should be able to proceed with the merge and such.</p>

**Comment by SimpleSamples** [comment]
<p>Thank you. I think it is fixed.</p>
<p>I used:</p>
<pre><code>git pull --allow-unrelated-histories
</code></pre>
<p>as in <a href="https://github.community/t/how-to-deal-with-refusing-to-merge-unrelated-histories-error/1372">Solved: How to deal with "refusing to merge unrelated hist… - GitHub Community Forum</a>. Then I did the push and now everything seems to be (“uploaded”) pushed. I do try to use proper terminology but in this situation I decided to use common terminology because some people really do not appreciate use of the wrong terminology.</p>

**Comment by mpboom** [comment]
<p>That is great news, thank you for sharing your journey with the community <a class="mention" href="/simplesamples">@simplesamples</a> !</p>
